File: rust/tests/adapter_migration_test.rs
//! AV-09–AV-11: mock v1.1 adapter migration tests.
//!
//! Mirrors `python/tests/test_adapter_migration.py`'s adapter-level tests
//! (test_v1_1_read, test_v1_1_roundtrip, test_v1_to_v1_1, test_v1_1_to_v1,
//! test_v1_unaffected) and `nodejs/tests/adapterMigration.test.ts`.
//!
//! Deliberately missing, by design (see VALIDATION_PLAN.md §9.3 and
//! `mock_v1_1/mod.rs`'s module doc): Python's/Node's *dispatcher-level*
//! tests (`test_dispatcher_v1_to_v1_1`, `test_dispatcher_v1_1_to_v1`), which
//! prove the public `MachineConfigReader`/`Writer` facade itself routes to
//! the mock via a temporarily-injected dispatch-table entry. Rust's public
//! dispatcher is a hardcoded `match "1.0" => ...` in `reader.rs`/`writer.rs`,
//! not a registry — there's no entry to inject. AV-09's actual rationale
//! ("adding v1.1 doesn't require modifying the v1.0 adapter") is satisfied
//! here by the mock living in its own file with zero edits to
//! `capabilities/v1_0/`, plus the full pre-existing suite (run below via
//! `v1_unaffected`, and via every other test in this crate) staying green.
//! Also skipped: Python's `test_mock_adapters_satisfy_protocol` — Rust has
//! no adapter trait/protocol to satisfy (the `adapters/` module was removed
//! as dead code earlier this branch).

mod mock_v1_1;

use machine_config::{MachineConfigReader, MachineConfigWriter};
use mock_v1_1::{make_mock_config, MockV1_1Reader, MockV1_1Writer};
use tempfile::NamedTempFile;

static REFERENCE: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/../fixtures/reference_config.h5");

/// Mock v1.1 file → StableModel — all five change categories asserted.
#[test]
fn v1_1_read_all_categories() {
    let cfg = make_mock_config(Some("MigrationTestMachine"), Some("Lab-001"), Some("TestEngineer"));
    let p = NamedTempFile::new().unwrap();
    MockV1_1Writer::new(&cfg).write(p.path()).unwrap();
    let result = MockV1_1Reader::new(p.path()).parse().unwrap();

    // ADDITION (x2)
    assert_eq!(result.meta.facility_id, Some("Lab-001".to_string()));
    assert_eq!(result.meta.config_author, Some("TestEngineer".to_string()));

    // REMOVAL (x2)
    assert_eq!(result.machine.gas_flow_direction, None);
    assert_eq!(result.machine.recoat_direction, None);

    // NAME (x2)
    assert_eq!(result.machine.machine_name, "MigrationTestMachine");
    assert_eq!(result.optical_trains[0].scanner.working_distance, cfg.optical_trains[0].scanner.working_distance);

    // PATH (x2)
    assert_eq!(result.machine.build_plate_z, cfg.machine.build_plate_z);
    assert_eq!(result.machine.build_plate_radius, cfg.machine.build_plate_radius);

    // NAME+PATH (x2)
    assert_eq!(result.machine.build_plate_x, cfg.machine.build_plate_x);
    assert_eq!(result.machine.build_plate_y, cfg.machine.build_plate_y);
}

/// Mock v1.1 → StableModel → mock v1.1 → StableModel — all categories
/// survive both passes.
#[test]
fn v1_1_roundtrip() {
    let cfg = make_mock_config(None, Some("RoundtripLab"), Some("RoundtripEngineer"));
    let p1 = NamedTempFile::new().unwrap();
    MockV1_1Writer::new(&cfg).write(p1.path()).unwrap();
    let mid = MockV1_1Reader::new(p1.path()).parse().unwrap();

    let p2 = NamedTempFile::new().unwrap();
    MockV1_1Writer::new(&mid).write(p2.path()).unwrap();
    let result = MockV1_1Reader::new(p2.path()).parse().unwrap();

    assert_eq!(result.meta.facility_id, Some("RoundtripLab".to_string()));
    assert_eq!(result.meta.config_author, Some("RoundtripEngineer".to_string()));
    assert_eq!(result.machine.gas_flow_direction, None);
    assert_eq!(result.machine.recoat_direction, None);
    assert_eq!(result.machine.machine_name, cfg.machine.machine_name);
    assert_eq!(
        result.optical_trains[0].scanner.working_distance,
        cfg.optical_trains[0].scanner.working_distance
    );
    assert_eq!(result.machine.build_plate_z, cfg.machine.build_plate_z);
    assert_eq!(result.machine.build_plate_radius, cfg.machine.build_plate_radius);
    assert_eq!(result.machine.build_plate_x, cfg.machine.build_plate_x);
    assert_eq!(result.machine.build_plate_y, cfg.machine.build_plate_y);
}

/// Real v1.0 fixture → StableModel → mock v1.1 layout — surviving fields
/// preserved; ADDITION fields None (no v1.0 source).
#[test]
fn v1_to_v1_1_forward_migration() {
    let source = MachineConfigReader::open(REFERENCE).unwrap().parse().unwrap();
    let mut migrated_input = source.clone();
    migrated_input.meta.file_version = "1.1-mock".to_string();

    let out = NamedTempFile::new().unwrap();
    MockV1_1Writer::new(&migrated_input).write(out.path()).unwrap();
    let result = MockV1_1Reader::new(out.path()).parse().unwrap();

    // NAME
    assert_eq!(result.machine.machine_name, source.machine.machine_name);
    // NAME+PATH
    assert_eq!(result.machine.build_plate_x, source.machine.build_plate_x);
    assert_eq!(result.machine.build_plate_y, source.machine.build_plate_y);
    // PATH
    assert_eq!(result.machine.build_plate_z, source.machine.build_plate_z);
    assert_eq!(result.machine.build_plate_radius, source.machine.build_plate_radius);
    // NAME (scanner)
    assert_eq!(
        result.optical_trains[0].scanner.working_distance,
        source.optical_trains[0].scanner.working_distance
    );
    // REMOVAL: always None regardless of what the v1.0 source contained
    assert_eq!(result.machine.gas_flow_direction, None);
    assert_eq!(result.machine.recoat_direction, None);
    // ADDITION: no v1.0 source -> typed fields None after forward migration
    assert_eq!(result.meta.facility_id, None);
    assert_eq!(result.meta.config_author, None);
}

/// Mock v1.1 file → StableModel → v1.0 layout — surviving fields preserved;
/// ADDITION fields lost (v1.0 writer doesn't write them).
#[test]
fn v1_1_to_v1_backward_migration() {
    let cfg = make_mock_config(None, Some("Lab-V11"), Some("MigrationBot"));

    let v1_1_file = NamedTempFile::new().unwrap();
    MockV1_1Writer::new(&cfg).write(v1_1_file.path()).unwrap();
    let v1_1_config = MockV1_1Reader::new(v1_1_file.path()).parse().unwrap();

    let mut downgrade_input = v1_1_config.clone();
    downgrade_input.meta.file_version = "1.0".to_string();

    let v1_out = NamedTempFile::new().unwrap();
    MachineConfigWriter::new(&downgrade_input).write(v1_out.path()).unwrap();
    let result = MachineConfigReader::open(v1_out.path()).unwrap().parse().unwrap();

    assert_eq!(result.machine.machine_name, cfg.machine.machine_name);
    assert_eq!(result.machine.build_plate_x, cfg.machine.build_plate_x);
    assert_eq!(result.machine.build_plate_y, cfg.machine.build_plate_y);
    assert_eq!(result.machine.build_plate_z, cfg.machine.build_plate_z);
    assert_eq!(result.machine.build_plate_radius, cfg.machine.build_plate_radius);
    assert_eq!(
        result.optical_trains[0].scanner.working_distance,
        cfg.optical_trains[0].scanner.working_distance
    );

    // REMOVAL: absent in v1.1 -> remain None after round-trip through v1.0
    assert_eq!(result.machine.gas_flow_direction, None);
    assert_eq!(result.machine.recoat_direction, None);

    // ADDITION: typed v1.1 fields are lost during backward migration
    assert_eq!(result.meta.facility_id, None);
    assert_eq!(result.meta.config_author, None);
}

/// Existing v1.0 read path is undisturbed — no mock adapter involved. This,
/// combined with every other test in this crate staying green after the
/// mock was added, is AV-09's actual proof: adding v1.1 required zero edits
/// to the v1.0 adapter.
#[test]
fn v1_unaffected() {
    let config = MachineConfigReader::open(REFERENCE).unwrap().parse().unwrap();
    assert_eq!(config.meta.file_version, "1.0");
    assert!(!config.optical_trains.is_empty());
    assert!(config.machine.build_plate_x.is_some());
}

File: nodejs/tests/adapterMigration.test.ts
/**
 * Adapter migration tests.
 *
 * Proves the StableModel contract holds across adapter version boundaries.
 * MockV1_1 (see `./mockV1_1.ts`) introduces all five change categories
 * relative to v1.0, exactly mirroring `python/tests/test_adapter_migration.py`
 * and `docs/migrations/mock_v1_0_to_v1_1.md`:
 *
 *   Addition  (x2): root attrs Facility_ID, Config_Author -> meta.facility_id/config_author
 *   Removal   (x2): Machine/ drops Gas_Flow_Direction, Recoat_Direction -> null
 *   Name      (x2): Machine/ "Machine_Name" -> "Machine_Label"
 *                   Scanner/ "Working_Distance" -> "Focal_Distance"
 *   Path      (x2): Build_Plate_Corner_Radius, Build_Plate_Z_Dimension
 *                   move from Machine/ to Machine/Dimensions/
 *   Name+path (x2): Build_Plate_X_Dimension -> Machine/Dimensions/ "Width"
 *                   Build_Plate_Y_Dimension -> Machine/Dimensions/ "Height"
 */
import { randomUUID } from 'node:crypto';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import type { MachineConfig } from '../src/models.js';
import { Hdf5AdapterV1_0, type ReadOptions } from '../src/capabilities/v1_0/hdf5.js';
import { Hdf5WriterV1_0 } from '../src/capabilities/v1_0/writer.js';
import { MachineConfigReader, _READERS } from '../src/reader.js';
import { MachineConfigWriter, _WRITERS } from '../src/writer.js';
import { FILE_VERSION, MockV1_1Reader, MockV1_1Writer, makeMockConfig } from './mockV1_1.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const REFERENCE = join(__dirname, '../../fixtures/reference_config.h5');

function tempH5Path(prefix: string): string {
  return join(tmpdir(), `${prefix}_${randomUUID()}.h5`);
}

// ---------------------------------------------------------------------------
// Tests — adapter-level (bypass the dispatcher)
// ---------------------------------------------------------------------------

describe('adapter migration — mock v1.1 (all five change categories)', () => {
  it('reads all five change categories from a mock v1.1 file', async () => {
    const cfg = makeMockConfig({ facilityId: 'Lab-001', configAuthor: 'TestEngineer' });
    const p = tempH5Path('v1_1_read');
    await new MockV1_1Writer(cfg).write(p);
    const result = await new MockV1_1Reader(p).parse();

    // ADDITION (x2)
    expect(result.meta.facility_id).toBe('Lab-001');
    expect(result.meta.config_author).toBe('TestEngineer');
    // REMOVAL (x2)
    expect(result.machine.gas_flow_direction).toBeNull();
    expect(result.machine.recoat_direction).toBeNull();
    // NAME (x2)
    expect(result.machine.machine_name).toBe('MigrationTestMachine');
    expect(result.optical_trains[0].scanner.working_distance).toBe(cfg.optical_trains[0].scanner.working_distance);
    // PATH (x2)
    expect(result.machine.build_plate_z).toBe(cfg.machine.build_plate_z);
    expect(result.machine.build_plate_radius).toBe(cfg.machine.build_plate_radius);
    // NAME+PATH (x2)
    expect(result.machine.build_plate_x).toBe(250.0);
    expect(result.machine.build_plate_y).toBe(175.0);
  });

  it('survives write -> read -> write -> read (roundtrip)', async () => {
    const cfg = makeMockConfig({ facilityId: 'RoundtripLab', configAuthor: 'RoundtripEngineer' });
    const p1 = tempH5Path('v1_1_a');
    await new MockV1_1Writer(cfg).write(p1);
    const mid = await new MockV1_1Reader(p1).parse();
    const p2 = tempH5Path('v1_1_b');
    await new MockV1_1Writer(mid).write(p2);
    const result = await new MockV1_1Reader(p2).parse();

    expect(result.meta.facility_id).toBe('RoundtripLab');
    expect(result.meta.config_author).toBe('RoundtripEngineer');
    expect(result.machine.gas_flow_direction).toBeNull();
    expect(result.machine.recoat_direction).toBeNull();
    expect(result.machine.machine_name).toBe(cfg.machine.machine_name);
    expect(result.optical_trains[0].scanner.working_distance).toBe(cfg.optical_trains[0].scanner.working_distance);
    expect(result.machine.build_plate_z).toBe(cfg.machine.build_plate_z);
    expect(result.machine.build_plate_radius).toBe(cfg.machine.build_plate_radius);
    expect(result.machine.build_plate_x).toBe(cfg.machine.build_plate_x);
    expect(result.machine.build_plate_y).toBe(cfg.machine.build_plate_y);
  });

  it('migrates a real v1.0 fixture forward to mock v1.1 — surviving fields preserved', async () => {
    const source = await new Hdf5AdapterV1_0(REFERENCE).parse();
    const forMock = { ...source, meta: { ...source.meta, file_version: FILE_VERSION } };
    const out = tempH5Path('migrated_v1_1');
    await new MockV1_1Writer(forMock).write(out);
    const result = await new MockV1_1Reader(out).parse();

    // NAME: machine_name survives both the read and write name remapping
    expect(result.machine.machine_name).toBe(source.machine.machine_name);
    // NAME+PATH: x and y survive via the StableModel intermediary
    expect(result.machine.build_plate_x).toBe(source.machine.build_plate_x);
    expect(result.machine.build_plate_y).toBe(source.machine.build_plate_y);
    // PATH: z and radius survive
    expect(result.machine.build_plate_z).toBe(source.machine.build_plate_z);
    expect(result.machine.build_plate_radius).toBe(source.machine.build_plate_radius);
    // NAME: scanner working_distance survives
    expect(result.optical_trains[0].scanner.working_distance).toBe(source.optical_trains[0].scanner.working_distance);
    // REMOVAL: always null regardless of what the v1.0 source contained
    expect(result.machine.gas_flow_direction).toBeNull();
    expect(result.machine.recoat_direction).toBeNull();
    // ADDITION: no v1.0 source -> fields are null after forward migration
    expect(result.meta.facility_id).toBeNull();
    expect(result.meta.config_author).toBeNull();
  });

  it('migrates a mock v1.1 file backward to v1.0 — ADDITION fields lost', async () => {
    const cfg = makeMockConfig({ facilityId: 'Lab-V11', configAuthor: 'MigrationBot' });
    const v1_1Path = tempH5Path('source_v1_1');
    await new MockV1_1Writer(cfg).write(v1_1Path);
    const v1_1Config = await new MockV1_1Reader(v1_1Path).parse();

    const forV1 = { ...v1_1Config, meta: { ...v1_1Config.meta, file_version: '1.0' } };
    const v1Out = tempH5Path('migrated_v1');
    await new Hdf5WriterV1_0(forV1).write(v1Out);
    const result = await new Hdf5AdapterV1_0(v1Out).parse();

    expect(result.machine.machine_name).toBe(cfg.machine.machine_name);
    expect(result.machine.build_plate_x).toBe(cfg.machine.build_plate_x);
    expect(result.machine.build_plate_y).toBe(cfg.machine.build_plate_y);
    expect(result.machine.build_plate_z).toBe(cfg.machine.build_plate_z);
    expect(result.machine.build_plate_radius).toBe(cfg.machine.build_plate_radius);
    expect(result.optical_trains[0].scanner.working_distance).toBe(cfg.optical_trains[0].scanner.working_distance);

    // REMOVAL: absent in mock v1.1 -> remain null after roundtrip through v1.0
    expect(result.machine.gas_flow_direction).toBeNull();
    expect(result.machine.recoat_direction).toBeNull();
    // ADDITION: typed v1.1 fields are lost on downgrade (v1.0 writer does not write them)
    expect(result.meta.facility_id).toBeUndefined();
    expect(result.meta.config_author).toBeUndefined();
  });

  it('leaves the existing v1.0 read path completely undisturbed', async () => {
    const config = await new Hdf5AdapterV1_0(REFERENCE).parse();
    expect(config.meta.file_version).toBe('1.0');
    expect(config.optical_trains.length).toBeGreaterThan(0);
    expect(config.machine).toBeDefined();
    expect(config.machine.build_plate_x).not.toBeNull();
  });
});

// ---------------------------------------------------------------------------
// Tests — dispatcher-level (full public API via injected _READERS/_WRITERS)
// ---------------------------------------------------------------------------

describe('adapter migration — public API dispatch', () => {
  beforeEach(() => {
    _READERS[FILE_VERSION] = MockV1_1Reader as unknown as (typeof _READERS)[string];
    _WRITERS[FILE_VERSION] = MockV1_1Writer as unknown as (typeof _WRITERS)[string];
  });

  afterEach(() => {
    delete _READERS[FILE_VERSION];
    delete _WRITERS[FILE_VERSION];
  });

  it('routes a mock v1.1 file through the real MachineConfigReader/Writer, then migrates it to v1.0', async () => {
    const cfg = makeMockConfig({ machineName: 'DispatcherTest', facilityId: 'Dispatch-Lab' });
    const v1_1File = tempH5Path('dispatcher_v1_1');
    await new MockV1_1Writer(cfg).write(v1_1File);

    // Public API read — dispatcher peeks "1.1-mock" and routes to MockV1_1Reader
    const result = await new MachineConfigReader(v1_1File).parse();
    expect(result.meta.file_version).toBe(FILE_VERSION);
    expect(result.meta.facility_id).toBe('Dispatch-Lab');
    expect(result.machine.machine_name).toBe(cfg.machine.machine_name);
    expect(result.machine.build_plate_x).toBe(cfg.machine.build_plate_x);

    // Public API write to v1.0 — dispatcher sees file_version "1.0" and routes to Hdf5WriterV1_0
    const v1Out = tempH5Path('dispatcher_migrated_v1');
    await new MachineConfigWriter({ ...result, meta: { ...result.meta, file_version: '1.0' } }).write(v1Out);
    const final = await new MachineConfigReader(v1Out).parse();
    expect(final.meta.file_version).toBe('1.0');
    expect(final.machine.machine_name).toBe(cfg.machine.machine_name);
    expect(final.machine.build_plate_x).toBe(cfg.machine.build_plate_x);
  });

  it('routes a real v1.0 file migrated to mock v1.1 through the public API', async () => {
    const source = await new Hdf5AdapterV1_0(REFERENCE).parse();

    // Public API write — dispatcher sees file_version "1.1-mock" and routes to MockV1_1Writer
    const v1_1Out = tempH5Path('dispatcher_v1_1_out');
    await new MachineConfigWriter({ ...source, meta: { ...source.meta, file_version: FILE_VERSION } }).write(v1_1Out);

    // Public API read — dispatcher peeks "1.1-mock" and routes to MockV1_1Reader
    const result = await new MachineConfigReader(v1_1Out).parse();
    expect(result.meta.file_version).toBe(FILE_VERSION);
    expect(result.machine.machine_name).toBe(source.machine.machine_name);
    expect(result.machine.build_plate_x).toBe(source.machine.build_plate_x);
    expect(result.optical_trains[0].scanner.working_distance).toBe(source.optical_trains[0].scanner.working_distance);
    expect(result.machine.gas_flow_direction).toBeNull(); // removed in v1.1-mock
    expect(result.machine.recoat_direction).toBeNull();
  });

  it('satisfies the same structural shape MachineConfigReader/Writer expect', () => {
    // Compile-time structural check: if this assigns without error, MockV1_1Reader/Writer
    // are structurally compatible with what the dispatch tables require.
    const readerCtor: new (path: string) => { parse: (o?: ReadOptions) => Promise<MachineConfig> } = MockV1_1Reader;
    const writerCtor: new (config: MachineConfig) => { write: (path: string) => Promise<void> } = MockV1_1Writer;
    expect(readerCtor).toBe(MockV1_1Reader);
    expect(writerCtor).toBe(MockV1_1Writer);
  });
});
